File: problem-solving/_lc-0312-hard-burst-balloons/1-top-down.py
"""
https://leetcode.com/problems/burst-balloons/

Complexity
----------
Time: O(n^3)
Space: O(n^2)

Reference
---------
Explanation: https://leetcode.com/problems/burst-balloons/solutions/892552/for-those-who-are-not-able-to-understand-any-solution-with-diagram/
Code: https://leetcode.com/problems/burst-balloons/solutions/76228/share-some-analysis-and-explanations/
"""

class Solution:
    def maxCoins(self, nums: List[int]) -> int:
        nums = [1] + nums + [1]  # To handle edge cases
        cache = {}
        def coins(l, r):
            if l > r:
                return 0
            if (l, r) in cache:
                return cache[(l, r)]
            cache[(l, r)] = 0
            for m in range(l + 1, r):
                cache[(l, r)] = max(
                    cache[(l, r)],
                    nums[l]*nums[m]*nums[r] + coins(l, m) + coins(m, r))
            return cache[(l, r)]
        return coins(0, len(nums) - 1)

File: problem-solving/lc-312-hard-burst-balloons/1-top-down.go
/*
https://leetcode.com/problems/burst-balloons/

Complexity
----------
Time: O(n^3)
Space: O(n^2)

Reference
---------
Explanation: https://leetcode.com/problems/burst-balloons/solutions/892552/for-those-who-are-not-able-to-understand-any-solution-with-diagram/
Code: https://leetcode.com/problems/burst-balloons/solutions/76228/share-some-analysis-and-explanations/

NOTE: This gives TLE in golang although got submitted in python but slow.
*/

func max(x, y int) int {
    if x > y {
        return x
    }
    return y
}

func maxCoins(nums []int) int {
    N := make([]int, len(nums) + 2)
    copy(N[1:], nums)
    N[0], N[len(N) - 1] = 1, 1
    cache := make(map[string]int)
    var F func(int, int) int
    F = func(l, r int) int {
        if l > r {
            return 0
        }
        key := fmt.Sprintf("%d-%d", l, r)
        if v, ok := cache[key]; ok {
            return v
        }
        for m := l + 1; m < r; m++ {
            cache[key] = max(
                cache[key],
                N[l]*N[m]*N[r] + F(l, m) + F(m, r))
        }
        return cache[key]
    }
    return F(0, len(N) - 1)
}
